src/libraries/System.Private.CoreLib/src/System/Threading/ThreadPoolWorkQueue.cs

```diff
@@ -725,7 +725,12 @@ internal static bool Dispatch()
                     //
                     int currentTickCount = Environment.TickCount;
                     if (!ThreadPool.NotifyWorkItemComplete(threadLocalCompletionCountObject, currentTickCount))
+                    {
+                        // This thread is being parked and may remain inactive for a while. Transfer any thread-local work items
+                        // to ensure that they would not be heavily delayed.
+                        tl.TransferLocalWork();
```

Comment:
Also want to do this in `NotifyThreadBlocked`?

Reply:
I'm trying to imagine a somewhat realistic case where a thread would cause a local work item to be queued to itself and then wait on a task. `Task.Run().Wait()` or similar from a thread pool thread would create that situation, and I can see that in a situation where the global queue is also consistently loaded no work stealing would be happening to allow a `Task.Run().Wait()` to complete in reasonable time. If the `Task.Run()` intends to do sync work then it could be flagged as long-running. Maybe I'm missing another scenario. Is it realistic enough to worry about?

I hesitate a bit to do this because it's not cheap and it can inadvertently invert the priority of those local work items and I'm not sure if that could be bad in some case. Curious though.

Reply:
Other threads can always pick them up from workitem stealing, so probably not worth worrying about?

Reply:
Yea I'm mainly worried about the situation where the global queue is consistently loaded and rarely becomes empty. In those cases work stealing just wouldn't happen. So any local work items leftover in inactive threads may be heavily delayed. I made this particular change here mainly (but not only) because uncommon spikes of cooperative blocking can indeed cause some threads to remain inactive for quite a while. Prior to this, only hill climbing would reduce the thread count and that issue is not as bad because hill climbing also keeps upping and downing the thread count, and it's not very good at decreasing thread counts from a high number, so eventually those work items would run in somewhat reasonable time, though they can still be delayed a bit.

Reply:
CC @stephentoub in case he has any thoughts about this

Reply:
Thanks.  I'll review the PR on Tuesday. Want to go through it in detail.

Reply:
> Task.Run().Wait() or similar from a thread pool thread would create that situation, and I can see that in a situation where the global queue is also consistently loaded no work stealing would be happening to allow a Task.Run().Wait() to complete in reasonable time.

Task.Run(...).Wait() on a thread pool thread should only get to this notification code if another thread pool thread stole the task before the Wait call.  If the task is still in this thread's local queue, the Wait on it will dequeue it and run it as part of the Wait call, rather than blocking waiting for someone else to get it.

Maybe I'm misunderstanding the concern?

Reply:
Ah right, it doesn't seem like an issue to me then